src/MongoDB/ServerApi.c

```diff
@@ -0,0 +1,391 @@
+	if (is_set) {
+		ZVAL_BOOL(&strict, mongoc_optional_value(mongoc_server_api_get_strict(intern->server_api)));
+	} else {
+		ZVAL_NULL(&strict);
+	}
+
+	if (include_null || is_set) {
+		zend_hash_str_add(props, ZEND_STRL("strict"), &strict);
+	}
+
+	is_set = mongoc_optional_is_set(mongoc_server_api_get_deprecation_errors(intern->server_api));
+	if (is_set) {
+		ZVAL_BOOL(&deprecation_errors, mongoc_optional_value(mongoc_server_api_get_deprecation_errors(intern->server_api)));
+	} else {
+		ZVAL_NULL(&deprecation_errors);
+	}
+
+	if (include_null || is_set) {
+		zend_hash_str_add(props, ZEND_STRL("deprecationErrors"), &deprecation_errors);
+	}
+
+	return props;
+} /* }}} */
+
+/* {{{ proto array MongoDB\Driver\ServerApi::bsonSerialize()
+*/
+static PHP_METHOD(ServerApi, bsonSerialize)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	ZVAL_ARR(return_value, php_phongo_serverapi_get_properties_hash(PHONGO_COMPAT_OBJ_P(getThis()), true, false));
+	convert_to_object(return_value);
+} /* }}} */
+
+/* {{{ proto string MongoDB\Driver\ServerApi::serialize()
+*/
+static PHP_METHOD(ServerApi, serialize)
+{
+	php_phongo_serverapi_t* intern;
+	zval                    retval;
+	php_serialize_data_t    var_hash;
+	smart_str               buf = { 0 };
+
+	intern = Z_SERVERAPI_OBJ_P(getThis());
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	array_init_size(&retval, 3);
+
+	ADD_ASSOC_STRING(&retval, "version", mongoc_server_api_version_to_string(mongoc_server_api_get_version(intern->server_api)));
+
+	if (mongoc_optional_is_set(mongoc_server_api_get_strict(intern->server_api))) {
+		ADD_ASSOC_BOOL_EX(&retval, "strict", mongoc_optional_value(mongoc_server_api_get_strict(intern->server_api)));
+	} else {
+		ADD_ASSOC_NULL_EX(&retval, "strict");
+	}
+
+	if (mongoc_optional_is_set(mongoc_server_api_get_deprecation_errors(intern->server_api))) {
+		ADD_ASSOC_BOOL_EX(&retval, "deprecationErrors", mongoc_optional_value(mongoc_server_api_get_deprecation_errors(intern->server_api)));
+	} else {
+		ADD_ASSOC_NULL_EX(&retval, "deprecationErrors");
+	}
+
+	PHP_VAR_SERIALIZE_INIT(var_hash);
+	php_var_serialize(&buf, &retval, &var_hash);
+	smart_str_0(&buf);
+	PHP_VAR_SERIALIZE_DESTROY(var_hash);
+
+	PHONGO_RETVAL_SMART_STR(buf);
+
+	smart_str_free(&buf);
+	zval_ptr_dtor(&retval);
+} /* }}} */
+
+/* {{{ proto void MongoDB\Driver\ServerApi::unserialize(string $serialized)
+*/
+static PHP_METHOD(ServerApi, unserialize)
+{
+	zend_error_handling     error_handling;
+	php_phongo_serverapi_t* intern;
+	char*                   serialized;
+	size_t                  serialized_len;
+	zval                    props;
+	php_unserialize_data_t  var_hash;
+
+	intern = Z_SERVERAPI_OBJ_P(getThis());
+
+	zend_replace_error_handling(EH_THROW, phongo_exception_from_phongo_domain(PHONGO_ERROR_INVALID_ARGUMENT), &error_handling);
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_STRING(serialized, serialized_len)
+	ZEND_PARSE_PARAMETERS_END_EX(
+		zend_restore_error_handling(&error_handling);
+		return );
```

Comment:
Is the semicolon after `return` purposely omitted? I know the macro adds its own, but for our readability perhaps it makes sense to write this as:

```c
ZEND_PARSE_PARAMETERS_END_EX(
    zend_restore_error_handling(&error_handling);
    return;
);
```

Reply:
I omitted it on purpose since the macro appends its own, but after seeing what clang-format does when we do have a semicolon (the resulting line is `return;);` I decided to extract all this functionality into corresponding `PHONGO_PARSE_PARAMETERS` macros which take care of calling `zend_replace_error_handling` appropriately.